src/Functions/seriesDecomposeSTL.cpp

```diff
@@ -0,0 +1,183 @@
+#include "config.h"
+
+#if USE_SEASONAL
+#    ifdef __clang__
+#        pragma clang diagnostic push
+#        pragma clang diagnostic ignored "-Wold-style-cast"
+#        pragma clang diagnostic ignored "-Wshadow"
+#        pragma clang diagnostic ignored "-Wimplicit-float-conversion"
+#    endif
+
+#    include <stl.hpp>
+
+#    ifdef __clang__
+#        pragma clang diagnostic pop
+#    endif
+
+#    include <cmath>
+#    include <Columns/ColumnArray.h>
+#    include <Columns/ColumnConst.h>
+#    include <Columns/ColumnsNumber.h>
+#    include <DataTypes/DataTypeArray.h>
+#    include <DataTypes/DataTypesNumber.h>
+#    include <Functions/FunctionFactory.h>
+#    include <Functions/FunctionHelpers.h>
+#    include <Functions/IFunction.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int ILLEGAL_COLUMN;
+}
+
+/*
+Decompose time series data based on STL(Seasonal-Trend Decomposition Procedure Based on Loess)
+*/
+class FunctionSeriesDecomposeSTL : public IFunction
+{
+public:
+    static constexpr auto name = "seriesDecomposeSTL";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionSeriesDecomposeSTL>(); }
+
+    std::string getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 2; }
+
+    bool isVariadic() const override { return true; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"time-series", &isArray<IDataType>, nullptr, "Array"},
+            {"period", &isNumber<IDataType>, nullptr, "Number"},
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeArray>(std::make_shared<DataTypeFloat32>()));
```

Comment:
So if the input is an array of Float64, will the output be Float32-based? Thats a loss of precision.

Reply:
The `str-cpp` library currents supports upto float32 precision only.